### ISSUE TITLE
Consider adding type expression syntax to exclude or explicitly include subtypes

### ISSUE THREAD

**Issue by elprans** [edgedb/edgedb#1605]
On queries over type `Foo` EdgeDB will include both `Foo` objects and all objects of its descendant types.  EdgeQL currently lacks a way to turn off automatic inclusion of inherited objects (a-la the SQL `ONLY` keyword).  There is a way to filter over the exact value of `.__type__`, but that is neither elegant, nor very efficient if spelled in a most concise way (`FILTER .__type__.name = 'default::BaseType').

`ONLY` seems to be a good candidate for a modifier in a particular case of excluding all descendants.

By nature the `ONLY` operator (and its tentative `EXCLUDING` counterpart) are type expression operators similar to the type union `|` and intersection `&`, so perhaps #1578 should be implemented first to avoid the tedium of repeating the inheritance op, e.g `WITH MyType AS TYPE Foo ONLY` or `WITH MyType AS TYPE Foo EXCLUDING Subtype`.  Also, perhaps `EXCLUDING` should be `-` or `^` to be closer to other type operators.

**Comment by 1st1** [comment]
> Also, perhaps EXCLUDING should be - or ^ to be closer to other type operators.

I don't think we should award relatively obscure type operations with single-character operators.

**Comment by elprans** [comment]
> I don't think we should award relatively obscure type operations with single-character operators.

Well, the type expression space isn't very crowded and is already a fairly advanced piece of functionality.

**Comment by vpetrovykh** [comment]
So if `ONLY` is a type expression operator then "select specifically Foo" would be `SELECT Foo[IS ONLY Foo]`, right?

**Comment by elprans** [comment]
> So if `ONLY` is a type expression operator then "select specifically Foo" would be `SELECT Foo[IS ONLY Foo]`, right?

Correct, although I'm thinking a postfix form would be better.

**Comment by 1st1** [comment]
> Well, the type expression space isn't very crowded and is already a fairly advanced piece of functionality.

Doesn't mean we should make it more crowded.

What's the use case for `EXCEPT` or `EXCLUDING` anyways? It sounds like it would be useful if you want to query by a base type, but we have `abstract` types and if they're used there's no need for `EXCEPT`.

And maybe we should spell out good use cases for `ONLY`.  Because if there are only one or two then `FILTER .__type__.name = 'default::BaseType'` approach works just fine and we don't need to add a new reserved keyword.

**Comment by elprans** [comment]
> `FILTER .__type__.name = 'default::BaseType'` approach works just fine

It works, but it can potentially be very slow.  Instead of statically excluding a type table from a `UNION`, you'd be joining a `Type` to the whole `UNION`.

>  add a new reserved keyword

It would be an unreserved keyword.

**Comment by 1st1** [comment]
> It would be an unreserved keyword.

Depends on where it would be placed. If it's similar to DETACHED it will similarly be a reserved one.  I'm not fond of `Foo[IS ONLY Foo]`.

**Comment by elprans** [comment]
> Depends on where it would be placed. 

`ONLY` is semantically a type expression operator, unlike `DETACHED`, which is an expression operator that manipulates _scope_.

> I'm not fond of `Foo[IS ONLY Foo]`.

Since `ONLY` is a type expression op, that's the only way to do this inline.  With aliasing this could be:

```
WITH OnlyFoo AS TYPE Foo ONLY
SELECT OnlyFoo;
```

**Comment by dmgolembiowski** [comment]
> > Depends on where it would be placed.
> 
> `ONLY` is semantically a type expression operator, unlike `DETACHED`, which is an expression operator that manipulates _scope_.

This is an interesting idea — especially since it fits the EdgeDB niche where "everything is a set". From my perspective, `[IS ONLY]` could be an implementation derived from creating disjoint unions, and could even play an interesting role with the `:-` operator.
